esp_isotp/src/esp_isotp.c

```diff
@@ -136,64 +381,214 @@ esp_err_t esp_isotp_new_transport(twai_node_handle_t twai_node, const esp_isotp_
+ *
+ * @param handle ISO-TP transport handle.
+ * @return ESP_OK on success or the error from TWAI disable.
+ */
+esp_err_t esp_isotp_delete(esp_isotp_handle_t handle)
+{
+    ESP_RETURN_ON_FALSE(handle, ESP_ERR_INVALID_ARG, TAG, "Invalid parameters");
+
+    esp_err_t ret = ESP_OK;
+
+    // Disable TWAI node after unregistering callbacks
+    esp_err_t twai_ret = twai_node_disable(handle->twai_node);
+    if (twai_ret != ESP_OK) {
+        ESP_LOGW(TAG, "Failed to disable TWAI node: %s", esp_err_to_name(twai_ret));
+        if (ret == ESP_OK) {
+            ret = twai_ret;
+        }
     }
-    free(isotp);
+
+    // Unregister TWAI callbacks first to prevent use-after-free during disable
+    twai_event_callbacks_t cbs = { 0 };
+    esp_err_t unreg_ret = twai_node_register_event_callbacks(handle->twai_node, &cbs, NULL);
+    if (unreg_ret != ESP_OK) {
+        ESP_LOGW(TAG, "Failed to unregister TWAI callbacks: %s", esp_err_to_name(unreg_ret));
+        ret = unreg_ret;
+    }
+
+    // Clean up ISO-TP link.
+    portENTER_CRITICAL(&handle->link_lock);
+    isotp_destroy_link(&handle->link);
+    portEXIT_CRITICAL(&handle->link_lock);
+
+    // Clean up TX frame array (SLIST pool is automatically cleaned when frames are freed).
+    if (handle->tx_frame_array) {
+        free(handle->tx_frame_array);
+    }
+
+    // Free ISO-TP reassembly buffers and handle.
+    if (handle->isotp_tx_buffer) {
+        free(handle->isotp_tx_buffer);
+    }
+    if (handle->isotp_rx_buffer) {
+        free(handle->isotp_rx_buffer);
+    }
+    free(handle);
+
     return ret;
 }
 
+
+/**
+ * @brief Poll the ISO-TP link. Call this periodically from a task.
+ *
+ * @param handle ISO-TP transport handle.
+ * @return ESP_OK on success, or an error code on failure.
+ */
 esp_err_t esp_isotp_poll(esp_isotp_handle_t handle)
 {
     ESP_RETURN_ON_FALSE(handle, ESP_ERR_INVALID_ARG, TAG, "Invalid parameters");
 
-    // Run ISO-TP state machine to check timeout and send continue frames
+    // Run ISO-TP state machine to check timeouts and send consecutive frames.
+    portENTER_CRITICAL(&handle->link_lock);
     isotp_poll(&handle->link);
+    portEXIT_CRITICAL(&handle->link_lock);
 
     return ESP_OK;
 }
 
+/**
+ * @brief Send a payload using ISO-TP.
+ *
+ * @param handle ISO-TP transport handle.
+ * @param data Pointer to payload buffer.
+ * @param size Payload size in bytes.
+ * @return
+ *  - ESP_OK on success
+ *  - ESP_ERR_NOT_FINISHED when the send is still in progress
+ *  - ESP_ERR_NO_MEM for buffer overflow conditions
+ *  - ESP_ERR_INVALID_SIZE for invalid sizes
+ *  - ESP_ERR_TIMEOUT on timeout
+ *  - ESP_FAIL for other errors
+ */
 esp_err_t esp_isotp_send(esp_isotp_handle_t handle, const uint8_t *data, uint32_t size)
 {
-    ESP_RETURN_ON_FALSE(handle && data && size, ESP_ERR_INVALID_ARG, TAG, "Invalid parameters");
+    ESP_RETURN_ON_FALSE_ISR(handle && data && size, ESP_ERR_INVALID_ARG, TAG, "Invalid parameters");
```

Comment:
if this function can also be called in ISR, we can skip the error print but just return the error code.